cudax/include/cuda/experimental/__utility/basic_any.cuh

```diff
@@ -0,0 +1,3146 @@
+inline constexpr __identity<_Ret (*)(void const*, _Args...) noexcept>
+  __virtual_override_fn<_Ret (*)(_Cp const&, _Args...) noexcept, _Tp, _Mbr> =
+    &__override_fn_<_Tp, _Mbr, _Ret, true, true, _Args...>;
+
+// TODO: Add support for member functions with reference qualifiers.
+
+template <class _Tp, auto _Mbr, class _Ret, class _Cp, class... _Args>
+inline constexpr __identity<_Ret (*)(void*, _Args...)> //
+  __virtual_override_fn<_Ret (_Cp::*)(_Args...), _Tp, _Mbr> = //
+  &__override_fn_<_Tp, _Mbr, _Ret, false, false, _Args...>;
+
+template <class _Tp, auto _Mbr, class _Ret, class _Cp, class... _Args>
+inline constexpr __identity<_Ret (*)(void const*, _Args...)> //
+  __virtual_override_fn<_Ret (_Cp::*)(_Args...) const, _Tp, _Mbr> =
+    &__override_fn_<_Tp, _Mbr, _Ret, true, false, _Args...>;
+
+template <class _Tp, auto _Mbr, class _Ret, class _Cp, class... _Args>
+inline constexpr __identity<_Ret (*)(void*, _Args...) noexcept>
+  __virtual_override_fn<_Ret (_Cp::*)(_Args...) noexcept, _Tp, _Mbr> =
+    &__override_fn_<_Tp, _Mbr, _Ret, false, true, _Args...>;
+
+template <class _Tp, auto _Mbr, class _Ret, class _Cp, class... _Args>
+inline constexpr __identity<_Ret (*)(void const*, _Args...) noexcept>
+  __virtual_override_fn<_Ret (_Cp::*)(_Args...) const noexcept, _Tp, _Mbr> =
+    &__override_fn_<_Tp, _Mbr, _Ret, true, true, _Args...>;
+
+template <class _Ret, class... _Args>
+_CUDAX_API _Ret __get_virtual_result(_Ret (*)(_Args...));
+
+template <class _Ret, class... _Args>
+_CUDAX_API _Ret __get_virtual_result(_Ret (*)(_Args...) noexcept) noexcept;
+
+template <class _Ret, class... _Args>
+_CUDAX_API _CUDA_VSTD::false_type __is_virtual_const(_Ret (*)(void*, _Args...));
+
+template <class _Ret, class... _Args>
+_CUDAX_API _CUDA_VSTD::true_type __is_virtual_const(_Ret (*)(void const*, _Args...));
+
+///
+/// __virtual_fn
+///
+template <auto _Fn>
+struct __virtual_fn
+{
+  using __function_t _CCCL_NODEBUG_ALIAS = decltype(__virtual_override_fn<decltype(_Fn)>);
+  using __result_t _CCCL_NODEBUG_ALIAS   = decltype(__get_virtual_result(__function_t{}));
+
+  static constexpr bool __const_fn   = decltype(__is_virtual_const(__function_t{}))::value;
+  static constexpr bool __nothrow_fn = noexcept(__get_virtual_result(__function_t{}));
+
+  template <class _Tp, auto _Mbr>
+  _CUDAX_API constexpr __virtual_fn(__override_t<_Tp, _Mbr>) noexcept
+      : __fn_(__virtual_override_fn<decltype(_Fn), _Tp, _Mbr>)
+  {}
+
+  __function_t __fn_;
+};
+
+///
+/// __ireference
+///
+/// Note: a `basic_any<__ireference<_Interface>>&&` is an rvalue reference, whereas
+/// a `basic_any<_Interface&>&&` is an lvalue reference.
+template <class _Interface>
+struct __ireference : _Interface
+{
+  static_assert(_CUDA_VSTD::is_class_v<_Interface>, "expected a class type");
+  static constexpr size_t __size_      = sizeof(void*);
+  static constexpr size_t __align_     = alignof(void*);
+  static constexpr bool __is_const_ref = _CUDA_VSTD::is_const_v<_Interface>;
```

Comment:
That is a bit misleading, we do not check whether it is a reference

Reply:
the type itself, `__ireference< I const >`, represents a const reference to an object that satisfies the `I` interface.